src/membership/members.rs

```diff
@@ -287,6 +308,38 @@ decl_module! {
             }
         }
 
+        pub fn set_controller_key(origin, controller: T::AccountId) {
+            let master = ensure_signed(origin)?;
+            let member_id = Self::ensure_is_member_primary_account(&master)?;
+
+            let mut profile = Self::ensure_profile(member_id)?;
+
+            // ensure new key is not used by someone else (master or controller)
+            ensure!(!<MemberIdByControllerAccountId<T>>::exists(&controller), "account already paired with member");
+            ensure!(!<MemberIdByAccountId<T>>::exists(&controller), "account already paired with member");
+
+            <MemberIdByControllerAccountId<T>>::remove(&profile.controller_account);
+            <MemberIdByControllerAccountId<T>>::insert(&controller, member_id);
+            profile.controller_account = controller.clone();
+            <MemberProfile<T>>::insert(member_id, profile);
+            Self::deposit_event(RawEvent::MemberSetControllerKey(member_id, controller));
+        }
+
+        pub fn set_primary_key(origin, new_primary: T::AccountId) {
+            let old_primary = ensure_signed(origin)?;
+            let member_id = Self::ensure_is_member_primary_account(&old_primary)?;
+
+            // ensure new key not is used by someone else (master or controller)
+            ensure!(!<MemberIdByControllerAccountId<T>>::exists(&new_primary), "account already paired with member");
```

Comment:
Don't check against the controller, even if controllers were forced to be unique would this check not be justified.

Reply:
That check is there to avoid needlessly updating the state if the new key is the same as the old. I will update the check to be clearer.